### ISSUE TITLE
[Feature] Option to select multiple collapsed parent tabs without children

### ISSUE THREAD

**Issue by Sada-Mio** [piroor/treestyletab#2557]
A (collapsed)
\ - A1 
B (collapsed)
\ - B1 
\ - B2
C (collapsed)
\ - C1

Hold CTRL + A,B,C or SHIFT + A,C

TST would select tabs including all childrens such as A1,B1,B2,C1. 

Would like option to only have parent tabs be select when tree collapsed.

**Comment by piroor** [comment]
Hmm, I'm negative to implement this behavior because it may make behavior less-user-friendly...

* When multiselected parent tabs are dragged, only those multiselected tabs will be moved and their non-selected descendants will be left there. This means that trees will be accidentally broken too easily. On the other hand, the current behavior (multiselecting collapsed tree selects all collapsed descendants too, so entire trees are moved together keeping their structure) looks more safe.
* Even if I implement such an option, I won't activate the behavior by default. But I'm afraid that people possibly activate accidentally, and it will introduce more troubles.
* "Multiselection of only parent tabs except children" is already doable on expanded tree. It is the regular way on TST: expanding trees at first and multiselecting them after that.

Basically I'm very negative to add options introducing more internal hidden (invisible) mode. I believe that any GUI should be more modeless. "Ctrl-click on a collapsed tree, then only the parent tab is selected but children are not, so children will be left there after I drag the selected parent tab" - such complicated behavior will enforce users to memorize and imagine internal mode of TST,  it is in the opposite of the modeless policy.

**Comment by Sada-Mio** [comment]
I see your point of keeping complexity straightforward both for user and dev end. 

Expanded tab selection for parents work but tedious with many tabs each with many descendants. Lots of scrolling and clicking. I just want only the parent tab to  be refreshed/reloaded. Perhaps an extension to TST would be a better approach.

**Comment by piroor** [comment]
OK, I've understood your motivation. Here is an example implementation to help such an usecase.

manifest.json:

```json
{
  "name": "Select Root Tabs",
  "version": "1",
  "manifest_version": 2,
  "background": {
    "scripts": ["background.js"]
  },
  "permissions": [
    "menus"
  ]
}
```

background.js:

```javascript
browser.menus.create({
  id:       'select-root-tabs',
  title:    'Select Root Tabs',
  contexts: ['tab']
});
browser.menus.create({
  id:       'select-root-parent-tabs',
  title:    'Select Root Tabs having any child',
  contexts: ['tab']
});

browser.menus.onClicked.addListener(async (info, contextTab) => {
  const tabs       = await getMultiselectedTabs(contextTab);
  const rootTabs   = await extractRootTabs(tabs);
  const parentTabs = rootTabs.filter(tab => tab.children.length > 0);
  switch (info.menuItemId) {
    case 'select-root-tabs':
      if (rootTabs.length != tabs.length)
        browser.tabs.highlight({
          windowId: contextTab.windowId,
          tabs:     rootTabs.map(tab => tab.index)
        });
      break;
    case 'select-root-parent-tabs':
      if (parentTabs.length != tabs.length)
        browser.tabs.highlight({
          windowId: contextTab.windowId,
          tabs:     parentTabs.map(tab => tab.index)
        });
      break;
  }
});

browser.menus.onShown.addListener(async (info, contextTab) => {
  const tabs       = await getMultiselectedTabs(contextTab);
  const rootTabs   = await extractRootTabs(tabs);
  const parentTabs = rootTabs.filter(tab => tab.children.length > 0);
  browser.menus.update('select-root-tabs', {
    enabled: rootTabs.length != tabs.length
  });
  browser.menus.update('select-root-parent-tabs', {
    enabled: parentTabs.length != tabs.length
  });
  browser.menus.refresh();
});

async function extractRootTabs(tabs) {
  const tabIds = tabs.map(tab => tab.id);
  tabs = await browser.runtime.sendMessage('treestyletab@piro.sakura.ne.jp', {
    type: 'get-tree',
    tabs: tabIds
  });
  const tabIdsSet = new Set(tabIds);
  const rootTabs  = tabs.filter(tab => !tab.ancestorTabIds.some(id => tabIdsSet.has(id)));
  const activeTab = rootTabs.find(tab => tab.active);
  const activeTabAncestors = activeTab ? rootTabs.filter(tab => tab.ancestorTabIds.includes(activeTab.id)) : [];
  // We need to put the active tab at the top of the list, to keep tab focus.
  return Array.from(new Set(activeTab ? [activeTab, ...activeTabAncestors, ...rootTabs] : [...activeTabAncestors, ...rootTabs]));
}

async function getMultiselectedTabs(tab) {
  if (tab.highlighted)
    return browser.tabs.query({
      windowId:    tab.windowId,
      highlighted: true
    });
  else
    return [tab];
}
```

Currently I have no confirmation of necessity whether these features should be built-in of TST globally or not. So I put this snippet under the public domain and I hope it is maintained by your or someone...

**Comment by irvinm** [comment]
@piroor I vote we close this one as well given the potential workaround.

**Comment by piroor** [comment]
OK, I close this.

Because I have limited resources (time and motivation) for development, I cannot host all helper addons by me. Instead I host only some helper addons which have possibility I use. Such a decision is always very arbitrary, sorry...
